src/OrleansLoggerProvider/OrleansLogger.cs

```diff
@@ -0,0 +1,119 @@
+        /// For example, in order to use message bulking feature, one need to use eventId = OrleansLogger.CreateEventId(eventId, errorCode) to create a EventId which fulfils the certain format.
+        /// Or one can use extension method <see cref="ILogger.Log(this ILogger logger, int errorCode, Severity sev, string format, object[] args, Exception exception)"/> to achieve this.
+        /// </summary>
+        /// <typeparam name="TState"></typeparam>
+        /// <param name="logLevel"></param>
+        /// <param name="eventId"></param>
+        /// <param name="state"></param>
+        /// <param name="exception"></param>
+        /// <param name="formatter"></param>
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        {
+            var errorCode = eventId.Id;
+            var severity = LogLevelToSeverity(logLevel);
+            WriteLogMessageToLogConsumers(errorCode, severity, formatter(state, exception), exception);
+        }
+
+        /// <summary>
+        /// Map LogLevel to Severity
+        /// </summary>
+        /// <param name="logLevel"></param>
+        /// <returns></returns>
+        public static Severity LogLevelToSeverity(LogLevel logLevel)
+        {
+            switch (logLevel)
+            {
+                case LogLevel.None: return Severity.Off;
+                case LogLevel.Critical: return Severity.Error;
+                case LogLevel.Error: return Severity.Error;
+                case LogLevel.Warning: return Severity.Warning;
+                case LogLevel.Information: return Severity.Info;
+                case LogLevel.Debug: return Severity.Verbose;
+                default: return Severity.Verbose3;
+            }
+        }
+
+        private void WriteLogMessageToLogConsumers(int errorCode, Severity sev, string message, Exception exception)
+        {
+            foreach (ILogConsumer consumer in this.logConsumers)
+            {
+                try
+                {
+                    consumer.Log(sev, this.name, message, exception, errorCode);
+                }
+                catch (Exception exc)
+                {
+                    Console.WriteLine("Exception while passing a log message to log consumer. LogConsumer type:{0}, logger name:{1}, severity:{2}, message:{3}, error code:{4}, message exception:{5}, log consumer exception:{6}",
+                        consumer.GetType().FullName, this.name, sev, message, errorCode, exception, exc);
+                }
+            }
+
+            //flush flushable consumers
+            if ((DateTime.UtcNow - lastFlush) > flushInterval)
+            {
+                lastFlush = DateTime.UtcNow;
+                foreach (IFlushableLogConsumer consumer in this.logConsumers.OfType<IFlushableLogConsumer>())
```

Comment:
The `this.logConsumers.OfType<IFlushableLogConsumer>()` could be computed once and enumerated to a list in the ctor.